src/main/java/dk/kb/discover/util/solrshield/Param.java

```diff
@@ -0,0 +1,303 @@
+        super(profile, name);
+        if (profile == null) {
+            throw new NullPointerException("Profile was null, but must be set for Params");
+        }
+        this.name = name;
+        enabled = config.getBoolean("default_enabled", enabled);
+        allowed = config.getBoolean("allowed", allowed);
+        weightConstant = config.getDouble("weight_constant", weightConstant);
+        weightFactor = config.getDouble("weight_factor", weightFactor);
+        this.multiValue = multiValue;
+    }
+
+    /**
+     * Assign the given values to the Param.
+     * @param values array of values, represented as Strings.
+     */
+    public void apply(String[] values) {
+        if (values == null || values.length == 0) {
+            return;
+        }
+        if (!multiValue && values.length > 1) {
+            log.debug("Param {} is single value but got {} values. Only the first value is used",
+                    name, values.length);
+            // TODO: Flag as invalid
+        }
+        enabled = true;
+        applyTypes(values);
+    }
+
+    /**
+     * Assign the given values to the Param.
+     * <p>
+     * {@code values} are guaranteed to contain at least 1 element.
+     * <p>
+     * {@link #enabled} will be set to true before this method is called. If the concrete {@code values} require
+     * {@link #enabled} to be set to false, it is up to the method implementation to do so (e.g. for boolean switches).
+     * @param values array of values, represented as Strings.
+     */
+    protected abstract void applyTypes(String[] values);
+
+    public V getValue() {
+        return value;
+    }
+
+    @Override
+    double getWeight() {
+        return !enabled ? 0.0 : weightConstant;
+    }
+
+    @Override
+    public boolean isAllowed(List<String> reasons) {
+        if (enabled && !this.allowed) {
+            reasons.add("Param " + name + "=" + value + " not allowed as the param itself is not allowed");
+            return false;
+        }
+        return true;
+    }
+
+    /* ************************** Implementing classes below ************************** */
+
+    public static class StringParam extends Param<StringParam, String[]> {
+        public int maxChars = 2000;
+        // TODO: blacklist/whitelist regexp (filtering {!...} or /regexp/ from query
+
+        public StringParam(Profile profile, YAML config, boolean multiValue) {
+            super(profile, config, multiValue);
+            maxChars = config.getInteger("max_chars", maxChars);
+            if (config.containsKey("default_value")) {
+                value = new String[]{config.getString("default_value")};
+            }
+        }
+
+        @Override
+        protected void applyTypes(String[] values) {
+            value = multiValue ? values : new String[]{ values[0] };
+        }
+
+        @Override
+        double getWeight() {
+            return !enabled ? 0.0 :
+                    weightConstant + weightFactor*value.length;
```

Comment:
It was an attempt to be consistent with the other `getWeight()` methods and to ensure that the "short" part of the terniary is at the beginning, so that it does not get visually lost at the end. The `SearchComponent.getWeight()` is one of the places where this is more clear:
```Java
        return !enabled ? 0.0 :
                super.getWeight() +
                        q.getWeight() +
                        fq.getWeight() +
                        start.getWeight() +
                        rows.getWeight() + rows.getValue() * rows.weightFactor * fl.getWeight() +
                        qOp.getWeight() +
                        wt.getWeight() +
                        version.getWeight() +
                        indent.getWeight() +
                        debug.getWeight() +
                        (debug.enabled ? debugExplainStructured.getWeight() : 0.0);
```
If this is flipped it becomes 
```Java
        return enabled ?
                super.getWeight() +
                        q.getWeight() +
                        fq.getWeight() +
                        start.getWeight() +
                        rows.getWeight() + rows.getValue() * rows.weightFactor * fl.getWeight() +
                        qOp.getWeight() +
                        wt.getWeight() +
                        version.getWeight() +
                        indent.getWeight() +
                        debug.getWeight() +
                        (debug.enabled ? debugExplainStructured.getWeight() : 0.0) :
                0.0;
```
which I find more confusing to read.

Alternatively if can be wrapped in an `if`. Would that be better?
```Java
        if (enabled) {
            return super.getWeight() +
                    q.getWeight() +
                    fq.getWeight() +
                    start.getWeight() +
                    rows.getWeight() + rows.getValue() * rows.weightFactor * fl.getWeight() +
                    qOp.getWeight() +
                    wt.getWeight() +
                    version.getWeight() +
                    indent.getWeight() +
                    debug.getWeight() +
                    (debug.enabled ? debugExplainStructured.getWeight() : 0.0);
        }
        return 0.0;
```

Reply:
I like the consistency throughout the classes, Lets keep it as it is. I was just curios about your rationality.